returnn/tf/layers/rec.py

```diff
@@ -4986,6 +5016,59 @@ def _get_cheating_targets_and_src_beam_idxs(self, scores):
     src_beams = src_beams[:, 0]  # (batch,)
     return cheating_gold_targets, src_beams
 
+  def _enforce_prefixes(self, top_k_labels, all_scores, top_k_scores, batch_dim, beam_size):
+      """
+      This function replaces the target labels from beam search by the ones predefined by the target prefixes as long
+      as search is still at a position within the prefix. We also replace the scores such that they correspond to a
+      prediction of the prefixes.
+
+      :param tf.Tensor top_k_labels: target labels from beam seach, shape (batch * beam,)
+      :param tf.Tensor all_scores: scores before beam pruning, used to lookup prefix scores, shape (batch, beam, dim)
+      :param tf.Tensor top_k_scores: scores after beam pruning, shape (batch, beam)
+      :param tf.Tensor|int batch_dim: number of sequences in batch (without beam)
+      :param int beam_size: outgoing beam size of this layer
+      :return: labels (batch * beam,) and scores (batch, beam) of self.prefix_target as long as within prefix, after
+        that top_k_labels and top_k_scores from beam search
+      :rtype: (tf.Tensor, tf.Tensor)
+      """
+      assert self.prefix_target
+
+      # Get the labels of the prefixes which should be enforced. They are padded with zeros, therefore we will
+      # get zeros for those sequences where the current timestep is beyond the length of the prefix.
+      target_prefix_labels = self._get_target_value(
+        target=self.prefix_target).get_placeholder_as_batch_major()  # (batch * beam,), int32
+
+      # Get prefixes that have already ended (i.e. have a smaller length than the current time step).
+      target_prefix_ended = tf.equal(target_prefix_labels, 0)
```

Comment:
This has the hardcoded assumption that EOS=0. I would avoid that. Better rely on the target seq len (using rec step info to check the current pos).